File: frontend/src/constants/CRUDCodeGeneratorInput.ts
import {AllowedLanguages} from "./AllowedLanguages.tsx";
import {AllowedDBMS} from "./AllowedDBMS.tsx";
import {TypeEnum} from "./TypeEnum.ts";

export interface Validation {
    minLength?: number;
    maxLength?: number;
    pattern?: string;
    min?: number;
    max?: number;
    customCode?: string;
}


export interface FieldModel {
    name: string;
    type: TypeEnum;
    primaryKey?: boolean;
    autoIncrement?: boolean;
    nullable?: boolean;
    unique?: boolean;
    validations?: Validation;
}


export interface TableModel {
    name: string;
    fields: FieldModel[];
}

export interface ConnectionParameters {
    host: string;
    port: number;
    database_name: string;
}

export interface CRUDCodeGeneratorInput {
    table: TableModel;
    language: AllowedLanguages;
    dbms: AllowedDBMS;
    connectionParams: ConnectionParameters;
    customCode?: Record<string, string>;
}

export function createCRUDCodeGeneratorInput(
    table: TableModel,
    language: AllowedLanguages,
    dbms: AllowedDBMS,
    connectionParams: ConnectionParameters,
    customCode?: Record<string, string>
): CRUDCodeGeneratorInput {
    return {
        table,
        language,
        dbms,
        connectionParams,
        customCode,
    };
}

File: backend/core/crud/src/parsing/CRUDCodeGeneratorInput.py
from pydantic import BaseModel

from backend.core.crud.src.parsing.constants.allowed_dbms import AllowedDBMS
from backend.core.crud.src.parsing.constants.allowed_languages import AllowedLanguages
from backend.core.crud.src.parsing.input_elements.connection_parameters import ConnectionParameters
from backend.core.crud.src.parsing.input_elements.table_model import TableModel


# Define the model
class CRUDCodeGeneratorInput(BaseModel):
    table: TableModel
    language: AllowedLanguages
    dbms: AllowedDBMS
    connectionParams: ConnectionParameters
